webhook/resources/backuptarget/mutator.go

```diff
@@ -0,0 +1,67 @@
+package backuptarget
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	common "github.com/longhorn/longhorn-manager/webhook/common"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+)
+
+type backupTargetMutator struct {
+	admission.DefaultMutator
+	ds *datastore.DataStore
+}
+
+func NewMutator(ds *datastore.DataStore) admission.Mutator {
+	return &backupTargetMutator{ds: ds}
+}
+
+func (b *backupTargetMutator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "backuptargets",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.BackupTarget{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (b *backupTargetMutator) Create(request *admission.Request, newObj runtime.Object) (admission.PatchOps, error) {
+	return mutate(newObj)
+}
+
+func (b *backupTargetMutator) Update(request *admission.Request, oldObj runtime.Object, newObj runtime.Object) (admission.PatchOps, error) {
+	return mutate(newObj)
```

Comment:
If an user wants to remove the finalizer, `mutate()` function will always add it back.

Reply:
The finalizer should be only removed when trying to delete the backup target like backups ([backup mutator](https://github.com/longhorn/longhorn-manager/blob/6da14ea7d571c7e73b6eae663fc11154d346de1e/webhook/resources/backup/mutator.go#L91-L93)). Users can remove the finalizer when the `deletionTimeStamp` is not `nil` at [common.GetLonghornFinalizerPatchOpIfNeeded](https://github.com/longhorn/longhorn-manager/blob/6da14ea7d571c7e73b6eae663fc11154d346de1e/webhook/common/common.go#L28)